text/0000-temp-components-lib.md

```diff
@@ -0,0 +1,611 @@
+  type PropsWithChildren,
+  createContext,
+  useCallback,
+  useContext,
+  useEffect,
+  useMemo,
+  useReducer,
+} from 'react'
+import { createPortal } from 'react-dom'
+import type { ToastOptions } from '@pluralsight/headless-styles/types'
+
+const CLOSE_TOAST = 'close'
+const SHOW_TOAST = 'show'
+
+interface ToastState {
+  show: boolean,
+  heading?: string
+  description: string
+  onUndo?: () => {}
+}
+
+interface ToastAction {
+  type: CLOSE_TOAST | SHOW_TOAST
+  data: ToastState
+}
+
+function toastReducer(state: ToastState, action: ToastAction) {
+  switch (action.type) {
+    case SHOW_TOAST:
+      return {
+        ...state,
+        ...action.data
+      }
+
+    case CLOSE_TOAST:
+      return {
+        ...state,
+        sentiment: 'info',
+        heading: '',
+        description: '',
+        onUndo: undefined,
+        show: false
+      }
+
+    default:
+      throw new Error('Something very bad has happened. This should never occur.')
+  }
+}
+
+function showToastAction(dispatch: DispatchWithoutAction, data: ToastState) {
+  return dispatch({
+    type: SHOW_TOAST,
+    data,
+  })
+}
+
+function closeToastAction(dispatch: DispatchWithoutAction) {
+  return dispatch({
+    type: CLOSE_TOAST,
+    data: initialToastState
+  })
+}
+
+const initialToastState = {
+  show: false,
+  sentiment: 'info',
+  heading: '',
+  description: '',
+  onUndo: undefined,
+}
+
+const ToastContext = createContext<ToastState>(initialToastState)
+
+function ToastProvider(props: PropsWithChildren<{}>) {
+  const [state, dispatch] = useReducer<ToastState>(reducer, initialToastState)
+
+  const value = useMemo(
+    () => ({
+      state,
+      dispatch
+    }),
+    [state, dispatch]
+  )
+
+  return (
+    <ToastContext.Provider value={value}>
+      {props.children}
+
+      {state.show &&
+        createPortal(
+          <Toast
+            onClose={() => closeToastAction(dispatch)}
+            sentiment={state.sentiment}
+            onUndo={state.onUndo}
+          >
+            <ToastHeading>{state.heading}</ToastHeading>
+            <p>{state.description}</p>
+          </Toast>
+        , document.body)}
+    </ToastContext.Provider>
+  )
+}
+
+interface ToastHookOptions extends ToastOptions {
+  heading?: string
+  description: string
+  onUndo?: () => {}
+}
+
+export function useToast() {
+  const toastRef = useRef(null)
+  const context = useContext(ToastContext)
+
+  if (!context) {
+    throw new Error(
+      'useToast requires a ToastProvider to be used on a parent component.'
+    )
+  }
+
+  const handleShow = useCallback((options: ToastState) => {
+    showToastAction(dispatch, options)
+
+    toastRef.current = setTimeout(() => {
+      closeToastAction(dispatch)
+    }, [9000])
+  }, [dispatch])
+
+  useEffect(() => {
+    return () => clearTimeout(toastRef.current)
+  }, [])
+
+  return useMemo(() => ({
+    close: closeToastAction(context.dispatch)
+    show: handleShow
+  }), [context, handleShow])
+}
+```
+
```

Comment:
minor: Might want to wrap big examples like this in something like a `details` tag so they can be expanded/collapsed 🙂 

```suggestion
Assuming we start with the current [Toast recommended foundation](https://design.pluralsight.com/docs/next/reference/components/toast#basic-toast), we can simply create a Provider and hook
combo that looks something like the example below.

<details>

<summary>
**Toast Provider and hook example**
</summary>

<p>

```typescript
import {
  type DispatchWithoutAction,
  type PropsWithChildren,
  createContext,
  useCallback,
  useContext,
  useEffect,
  useMemo,
  useReducer,
} from 'react'
import { createPortal } from 'react-dom'
import type { ToastOptions } from '@pluralsight/headless-styles/types'

const CLOSE_TOAST = 'close'
const SHOW_TOAST = 'show'

interface ToastState {
  show: boolean,
  heading?: string
  description: string
  onUndo?: () => {}
}

interface ToastAction {
  type: CLOSE_TOAST | SHOW_TOAST
  data: ToastState
}

function toastReducer(state: ToastState, action: ToastAction) {
  switch (action.type) {
    case SHOW_TOAST:
      return {
        ...state,
        ...action.data
      }

    case CLOSE_TOAST:
      return {
        ...state,
        sentiment: 'info',
        heading: '',
        description: '',
        onUndo: undefined,
        show: false
      }

    default:
      throw new Error('Something very bad has happened. This should never occur.')
  }
}

function showToastAction(dispatch: DispatchWithoutAction, data: ToastState) {
  return dispatch({
    type: SHOW_TOAST,
    data,
  })
}

function closeToastAction(dispatch: DispatchWithoutAction) {
  return dispatch({
    type: CLOSE_TOAST,
    data: initialToastState
  })
}

const initialToastState = {
  show: false,
  sentiment: 'info',
  heading: '',
  description: '',
  onUndo: undefined,
}

const ToastContext = createContext<ToastState>(initialToastState)

function ToastProvider(props: PropsWithChildren<{}>) {
  const [state, dispatch] = useReducer<ToastState>(reducer, initialToastState)

  const value = useMemo(
    () => ({
      state,
      dispatch
    }),
    [state, dispatch]
  )

  return (
    <ToastContext.Provider value={value}>
      {props.children}

      {state.show &&
        createPortal(
          <Toast
            onClose={() => closeToastAction(dispatch)}
            sentiment={state.sentiment}
            onUndo={state.onUndo}
          >
            <ToastHeading>{state.heading}</ToastHeading>
            <p>{state.description}</p>
          </Toast>
        , document.body)}
    </ToastContext.Provider>
  )
}

interface ToastHookOptions extends ToastOptions {
  heading?: string
  description: string
  onUndo?: () => {}
}

export function useToast() {
  const toastRef = useRef(null)
  const context = useContext(ToastContext)

  if (!context) {
    throw new Error(
      'useToast requires a ToastProvider to be used on a parent component.'
    )
  }

  const handleShow = useCallback((options: ToastState) => {
    showToastAction(dispatch, options)

    toastRef.current = setTimeout(() => {
      closeToastAction(dispatch)
    }, [9000])
  }, [dispatch])

  useEffect(() => {
    return () => clearTimeout(toastRef.current)
  }, [])

  return useMemo(() => ({
    close: closeToastAction(context.dispatch)
    show: handleShow
  }), [context, handleShow])
}
```

</p>
</details>
```